pkg/vere/disk.c

```diff
@@ -867,66 +868,494 @@ u3_disk_init(c3_c* pax_c, u3_disk_cb cb_u)
+    }
+  }
+
+  //  get metadata from old epoch or unmigrated event log's db
+  c3_d     who_d[2];
+  c3_o     fak_o;
+  c3_w     lif_w;
+  if ( c3y == eps_o ) {
+    if ( c3y != u3_disk_read_meta(log_u->mdb_u, who_d, &fak_o, &lif_w) ) {
+      fprintf(stderr, "disk: failed to read metadata\r\n");
+      goto fail;
+    }
+  }
+
+  //  initialize db of new epoch
+  if ( c3y == u3_Host.ops_u.nuu || new_d > 0 ) {
+    c3_c dat_c[8193];
+    snprintf(dat_c, sizeof(dat_c), "%s/data.mdb", epo_c);
+    if ( 0 == (log_u->mdb_u = u3_lmdb_init(epo_c)) ) {
+      fprintf(stderr, "disk: failed to initialize database\r\n");
+      c3_free(log_u);
+      goto fail;
+    }
+  }
+
+  // write the metadata to the database
+  if ( c3y == eps_o ) {
+    if ( c3n == u3_disk_save_meta(log_u->mdb_u, who_d, fak_o, lif_w) ) {
+      fprintf(stderr, "disk: failed to save metadata\r\n");
+      goto fail;
+    }
+  }
+
+  //  load new epoch directory and set it in log_u
+  log_u->epo_u = u3_foil_folder(epo_c);
+
+  //  success
+  return c3y;
+
+fail:
+  c3_unlink(epv_c);
+  c3_unlink(biv_c);
+  c3_rmdir(epo_c);
+  return c3n;
+}
+
+/* u3_disk_epoc_kill: delete an epoch.
+*/
+c3_o u3_disk_epoc_kill(u3_disk* log_u, c3_d epo_d) {
+  //  get epoch directory
+  c3_c epo_c[8193];
+  snprintf(epo_c, sizeof(epo_c), "%s/0i%" PRIc3_d, log_u->com_u->pax_c, epo_d);
+
+  //  delete files in epoch directory
+  u3_dire* dir_u = u3_foil_folder(epo_c);
+  u3_dent* den_u = dir_u->all_u;
+  while ( den_u ) {
+    c3_c fil_c[8193];
+    snprintf(fil_c, sizeof(fil_c), "%s/%s", epo_c, den_u->nam_c);
+    if ( 0 != c3_unlink(fil_c) ) {
+      fprintf(stderr, "disk: failed to delete file in epoch directory\r\n");
+      return c3n;
+    }
+    den_u = den_u->nex_u;
+  }
+
+  //  delete epoch directory
+  if ( 0 != c3_rmdir(epo_c) ) {
+    fprintf(stderr, "disk: failed to delete epoch directory\r\n");
+    return c3n;
+  }
+
+  //  cleanup
+  u3_dire_free(dir_u);
+
+  //  success
+  return c3y;
+}
+
+/* u3_disk_epoc_last: get latest epoch number.
+*/
+c3_o u3_disk_epoc_last(u3_disk* log_u, c3_d* lat_d) {
+  c3_o ret_o = c3n;  //  return no if no epoch directories exist
+  *lat_d = 0;        //  initialize lat_d to 0
+  u3_dire* die_u = u3_foil_folder(log_u->com_u->pax_c);
+  u3_dent* den_u = die_u->dil_u;
+  while ( den_u ) {
+    c3_d epo_d = 0;
+    if ( 1 == sscanf(den_u->nam_c, "0i%" PRIc3_d, &epo_d) ) {
+      ret_o = c3y;   //  NB: returns yes if the directory merely exists
+      *lat_d = c3_max(epo_d, *lat_d);  //  update the latest epoch number
+    }
+    den_u = den_u->nex_u;
+  }
+  
+  u3_dire_free(die_u);
+
+  return ret_o;
+}
+
+/* u3_disk_epoc_vere: get binary version from epoch.
+*/
+c3_o
+u3_disk_epoc_vere(u3_disk* log_u, c3_d epo_d, c3_c* ver_w) {
+  //  XX  check implementation (similar code is in king.c, _king_get_pace)
+  //      should probably write a generic function to read a string from 
+  //      a file into malloc'd memory, with an optional whitespace trimming
```

Comment:
Bumping this comment, as this function signature is unsafe (`fil_u` is read into `ver_w`, regardless of length).